File: kalman/kalman.py
from math import *
import matplotlib.pyplot as plt
import numpy as np

#Gaussian implementation
def gaussian(mu, variance, x):
    return 1.0/sqrt(2.0*pi*variance) * exp(-0.5*(x-mu)**2/variance)

def update(mu1, var1, mu2, var2):
    mu = (var2*mu1 + var1*mu2)/(var1 + var2)
    var = 1/(1/var2 + 1/var1)
    return [mu,var]

def predict(mu1, var1, mu2, var2):
    mu = mu1 + mu2
    var = var1 + var2

    return [mu,var]

def current_gaussian(mu, var):
    print(f'Current Gaussian with mean-{mu} and variance-{var}')
    x_axis = np.arange(-30,30,0.1)
    g=[]
    for x in x_axis:
        g.append(gaussian(mu, var, x))

    plt.plot(x_axis, g)
    plt.show()

def main():
    measurements = [15.0,18.0,19.0,21.0,19.0]
    measurement_var = 4
    motion = [2.0,1.0,2.0,-2.0,1.0]
    motion_var = 2
    mu = 0
    var = 10000

    current_gaussian(mu, var)

    for i in range(len(measurements)):
        mu, var = update(mu, var, measurements[i], measurement_var)
        print(f'Update: [{mu}, {var}]')
        mu, var = predict(mu, var, motion[i], motion_var)
        print(f'Predict: [{mu}, {var}]')

    print(f'\nFinal position: [{mu}, {var}]')

    current_gaussian(mu, var)


if __name__=='__main__':
    main()

File: kalman/kalman.java
import java.util.*;

class Kalman{

    public Kalman(){
    }

    public double gaussian(double mu, double var, double x){
        return 1.0/Math.sqrt(2.0*3.14*var) * Math.exp(Math.pow(-0.5*(x-mu), 2)/var);
    }

    public ArrayList<Double> update(double mu1, double var1, double mu2, double var2){
        double mu = (var2*mu1 + var1*mu2)/(var1 + var2);
        double var = 1/(1/var2 + 1/var1);
        ArrayList<Double> arr = new ArrayList<Double>();
        arr.add(mu);
        arr.add(var);

        return arr;
    }

    public ArrayList<Double> predict(double mu1, double var1, double mu2, double var2){
        double mu = mu1 + mu2;
        double var = var1 + var2;
        ArrayList<Double> arr = new ArrayList<Double>();
        arr.add(mu);
        arr.add(var);

        return arr;
    }

    public static void main(String args[]) {
        Kalman k = new Kalman();
        Scanner s = new Scanner(System.in);
        System.out.println("Enter your no. of measurements: ");
        int num = s.nextInt();
        int num2 = num;
        System.out.println("Enter your measurements(space seperated): ");
        ArrayList<Double> measurements = new ArrayList<Double>();
        while(num-->0){
            measurements.add(s.nextDouble());
        }
        System.out.println("Enter your measurement variance: ");
        Double measurements_var = s.nextDouble();

        System.out.println("Enter your motions(space seperated): ");
        ArrayList<Double> motion = new ArrayList<Double>();
        while(num2-->0){
            motion.add(s.nextDouble());
        }
        if(motion.size()!=measurements.size()){
            System.out.println("Enter same no. of elements for measurements and motion");
            System.exit(0);
        }
        System.out.println("Enter your measurement variance: ");
        Double motion_var = s.nextDouble();
        Double mu = 0.0;
        Double var = 10000.0;

        for(int i=0;i<measurements.size();i++){
            ArrayList<Double> temp = k.update(mu, var, measurements.get(i), measurements_var);
            mu = temp.get(0);
            var = temp.get(1);
            System.out.println("Update: [" + mu + ", " + var + "]");
            ArrayList<Double> temp2 = k.predict(mu, var, motion.get(i), motion_var);
            mu = temp2.get(0);
            var = temp2.get(1);
            System.out.println("Predict: [" + mu + ", " + var + "]");
        }

        System.out.println("Final Position: [" + mu + ", " + var + "]");

        s.close();
    }

}